lib/counters/src/lib.rs

```diff
@@ -77,3 +78,132 @@ macro_rules! count {
+pub struct OptionCounters<T: Count> {
+    /// Counters for this [`Option`]'s [`Some`] variant.
+    pub Some: T::Counters,
+    /// The total number of [`None`]s recorded by this counter.
+    pub None: AtomicU32,
+}
+
+impl<T: Count, E: Count> Count for Result<T, E> {
+    type Counters = ResultCounters<T, E>;
+    const NEW_COUNTERS: Self::Counters = ResultCounters {
+        Ok: T::NEW_COUNTERS,
+        Err: E::NEW_COUNTERS,
+    };
+
+    fn count(&self, counters: &Self::Counters) {
+        match self {
+            Ok(t) => t.count(&counters.Ok),
+            Err(e) => e.count(&counters.Err),
+        }
+    }
+}
+
+impl<T: Count> Count for Option<T> {
+    type Counters = OptionCounters<T>;
+
+    #[allow(clippy::declare_interior_mutable_const)]
+    const NEW_COUNTERS: Self::Counters = OptionCounters {
+        Some: T::NEW_COUNTERS,
+        None: AtomicU32::new(0),
+    };
+
+    fn count(&self, counters: &Self::Counters) {
+        match self {
+            Some(t) => t.count(&counters.Some),
+            None => {
+                armv6m_atomic_hack::AtomicU32Ext::fetch_add(
+                    &counters.None,
+                    1,
+                    Ordering::Relaxed,
+                );
+            }
+        }
+    }
+}
+
+impl<T: Count, E: Count> Count for &'_ Result<T, E> {
+    type Counters = ResultCounters<T, E>;
+    const NEW_COUNTERS: Self::Counters = ResultCounters {
+        Ok: T::NEW_COUNTERS,
+        Err: E::NEW_COUNTERS,
+    };
+
+    fn count(&self, counters: &Self::Counters) {
+        match self {
+            Ok(t) => t.count(&counters.Ok),
+            Err(e) => e.count(&counters.Err),
+        }
+    }
+}
+
+impl<T: Count> Count for &'_ Option<T> {
+    type Counters = OptionCounters<T>;
+
+    #[allow(clippy::declare_interior_mutable_const)]
+    const NEW_COUNTERS: Self::Counters = OptionCounters {
+        Some: T::NEW_COUNTERS,
+        None: AtomicU32::new(0),
+    };
+
+    fn count(&self, counters: &Self::Counters) {
+        match self {
+            Some(t) => t.count(&counters.Some),
+            None => {
+                armv6m_atomic_hack::AtomicU32Ext::fetch_add(
+                    &counters.None,
+                    1,
+                    Ordering::Relaxed,
+                );
+            }
+        }
+    }
+}
+
+impl Count for core::convert::Infallible {
+    type Counters = AtomicU32;
+    #[allow(clippy::declare_interior_mutable_const)]
+    const NEW_COUNTERS: Self::Counters = AtomicU32::new(0);
+
+    fn count(&self, _counters: &Self::Counters) {
+        // `Infallible`s are not made. They should NEVER be made. We
+        // will not count them. We will not help count them.
+        #[cfg(debug_assertions)]
+        unreachable!();
+
+        // but just in case...
+        #[cfg(not(debug_assertions))]
+        armv6m_atomic_hack::AtomicU32Ext::fetch_add(
+            _counters,
+            1,
+            Ordering::Relaxed,
+        );
+    }
+}
+
+impl Count for () {
```

Comment:
This impl is kind of the first time where I've been a little concerned about us overflowing a U32. How often are we counting `()`?

Reply:
This impl is^H^Hwill be used for counting the `Ok(())` arm of a `Result<(), T> where T: Count`. In particular, for fallible IPC operations, Idol will generate an enum variant like 
```rust
enum ${name_of_ipc_interface}Event {
    ${name_of_fallible_ipc_op}(#[count(children)] Result<(), #{the_fallible_op_error_type}>,
    // ... other variants ...
}
```

so, there's no _global_ counter of `()`s; it's just used to instantiate a counter for the `Ok` side of a `Result` counter in a particular enum variant.

We could use similar counters in hand-written code; IIRC we have some counted ringbuffers that contain variants that are `Result<(), Error>` that aren't currently counting the `Ok`/`Err` variants separately, but could be. 

Does that explanation make sense?

Reply:
It does, thanks.